### PR TITLE
Gem build strict option

### LINKED ISSUE #1334: Allow treating warnings as errors during packaging / validation
This will allow for a 'no warnings' policy in an automated release pipeline, and would help prevent releasing broken gems

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
# Description:

This is a rework of #1335, applying (hopefully) all the suggestions made in there. There's also a little prior refactoring I needed to do in order to cleanly add the feature, but hopefully no breaking changes, at least not intentional... :)

@segiddins I started with a rebase of your work but due to heavy squashing and amending, credit was lost. Let me know if you'd like me to restore your authorship to 0bbb986 :wink:.

Fixes #1334.
Supersedes #1335.
______________

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by deivid-rodriguez** [comment]
Looking at the test failure, I think [this stub](https://github.com/bundler/bundler/blob/d44d803357506895555ff97f73e60d593820a0de/spec/support/builders.rb#L604) needs to be updated in `bundler`. How do you usually handle these situations?

**Comment by deivid-rodriguez** [comment]
I guess I need to add a compatible change to bundler to get it passing, wait for it to be released and wait for the new version to be vendored in this repo? Or is there a simpler/faster way?

**Comment by bronzdoc** [comment]
Is this PR and  https://github.com/rubygems/rubygems/pull/1335 trying to solve the same problem?

**Comment by deivid-rodriguez** [comment]
Yes, this PR is a continuation of #1335, rebased and with the feedback left in there addressed.

File: test/rubygems/test_gem_specification.rb [hunk 1]
```diff
@@ -2957,8 +2957,7 @@ def test_validate_license
     end
 
     assert_match <<-warning, @ui.error
-WARNING:  licenses is empty, but is recommended.  Use a license identifier from
-http://spdx.org/licenses or 'Nonstandard' for a nonstandard license.
+WARNING:  licenses is empty, but is recommended.  Use a license identifier from http://spdx.org/licenses or 'Nonstandard' for a nonstandard license.
```
Comment (segiddins) [inline test/rubygems/test_gem_specification.rb]: why has the wrapping on this changed?

Comment (deivid-rodriguez) [reply test/rubygems/test_gem_specification.rb]: I found the old wrapping weird looking and arbitrary. I found out about it when writing [this test](https://github.com/rubygems/rubygems/pull/2332/files#diff-7c25fba4c6ec502e025345c5f4d6a5edR89), where I was expecting to check for one warning per line until no more output, but I was getting the warning broken into two lines.

I agree it's probably a discussion for other PR, so I'll undo it!

Is history editing ok, by the way, or should I add the revert and test fixes as separate commits?

**Comment by deivid-rodriguez** [comment]
So the plan for getting this green would be:

* Add bundler/bundler#6601 to the `1.16.3` milestone.
* Wait for `bundler 1.16.3` to be released.
* Wait for `bundler 1.16.3` to be vendored in `rubygems`.
* Rebase this PR.

Right?

File: test/rubygems/test_gem_specification.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply test/rubygems/test_gem_specification.rb]: Reverted as a separate commit, I can cleanup history later.

**Comment by segiddins** [comment]
Sounds about right

**Comment by deivid-rodriguez** [comment]
Rebased now that bundler 1.16.3 has been vendored, it's green!

**Comment by bronzdoc** [comment]
@bundlerbot r+

**Pull request merged**